File: app/src/main/java/com/elvotra/clean/domain/executor/UseCaseScheduler.java
package com.elvotra.clean.domain.executor;

import com.elvotra.clean.domain.usecase.base.BaseUseCase;

public interface UseCaseScheduler {

    void execute(Runnable runnable);

    <V extends BaseUseCase.ResponseValue> void notifyResponse(final V response,
                                                              final BaseUseCase.UseCaseCallback<V> useCaseCallback);

    <V extends BaseUseCase.ResponseValue> void onError(int statusCode,
                                                       final BaseUseCase.UseCaseCallback<V> useCaseCallback);
}

File: app/src/main/java/com/elvotra/clean/domain/executor/UseCaseScheduler.kt
package com.elvotra.clean.domain.executor

import com.elvotra.clean.domain.usecase.base.BaseUseCase

interface UseCaseScheduler {

    fun execute(runnable: Runnable)

    fun <V : BaseUseCase.ResponseValue> notifyResponse(response: V,
                                                       useCaseCallback: BaseUseCase.UseCaseCallback<V>)

    fun <V : BaseUseCase.ResponseValue> onError(statusCode: Int,
                                                useCaseCallback: BaseUseCase.UseCaseCallback<V>)
}
